vignettes/NSSP_nowcast.Rmd

```diff
@@ -0,0 +1,539 @@
+```
+
+## Expand the data so that each "event" has its own column
+
+First we will pivot the line-list's time stamp and diagnosis update columns into a long format with one row per update.
+
+We will create two datasets which parse the characters in the columns `DischargeDiagnosisMDTUpdates` and `DischargeDiagnosisUpdates` , which contain a string listing the time stamp and diagnosis codes (respectively) of each "event" in the clinical encounter, formatted as:
+
+ - `{event number};YYYY-MM-DD HH:MM:SS;|{event number 2};YYYY-MM-DD HH:MM:SS;|` for `DischargeDiagnosisMDTUpdates`
+
+ - `{event number};{diagnoses codes};|{event number 2}{diagnoses codes};|` for `DischargeDiagnosisUpdates`
+
+The timestamp records the timing of diagnosis code updates related to each clinical encounter, capturing the point in time when each new code became available within the NSSP system.
+Later, we will merge the two datasets back together by the unique patient ID and the event number, so that we can associate each set of diagnoses codes with a timestamp.
+
+We will use `tidyr::separate_wider_delim()` to expand these entries, so that each "event" has its own column.
+Since patients experience a different number of patient update "events", there will be missing values for patients not experiencing many events during their visit.
+The columns will be named by the original column name + the event number, e.g. `DischargeDiagnosisMDTUpdates1`.
+```{r}
+syn_nssp_time_stamps_wide <- syn_nssp_line_list |>
+  separate_wider_delim(DischargeDiagnosisMDTUpdates,
+    delim = "{", names_sep = "", too_few = "align_start"
+  ) |>
+  select(-DischargeDiagnosisUpdates)
+head(syn_nssp_time_stamps_wide)
+```
+
+Repeat for the diagnoses codes column.
+```{r}
+syn_nssp_diagnoses_wide <- syn_nssp_line_list |>
+  separate_wider_delim(DischargeDiagnosisUpdates,
+    delim = "{", names_sep = "", too_few = "align_start"
+  ) |>
+  select(-DischargeDiagnosisMDTUpdates)
+```
+
+Find the name of the last update column, and pivot the data from wide to long.
+This creates a long tidy dataframe where each row is now an event.
+We will create a unique event ID using the event number and the patient ID in the `C_Processed_BioSense_ID` column.
+```{r}
+lastcol_ts <- tail(colnames(select(
+  syn_nssp_time_stamps_wide,
+  matches("DischargeDiagnosisMDTUpdates")
+)), 1)
+syn_nssp_time_stamps_long <- syn_nssp_time_stamps_wide |>
+  pivot_longer(
+    cols = starts_with("DischargeDiagnosisMDTUpdates"),
+    names_to = "column_name",
+    values_to = "time_stamp",
+    values_drop_na = FALSE
+  ) |>
+  mutate(
+    event_id = paste(
+      C_Processed_BioSense_ID,
+      parse_number(as.character(column_name))
+    )
+  )
+```
+
+Perform the same procedure for the diagnoses dataset.
+```{r}
+lastcol_diag <- tail(colnames(select(
+  syn_nssp_diagnoses_wide,
+  matches("DischargeDiagnosisUpdates")
+)), 1)
+syn_nssp_diagnoses_long <- syn_nssp_diagnoses_wide |>
```

Comment:
if this is literally identical code I would suggest it is more readable to define a function say why and then use it twice